filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilderFactory.scala

```diff
@@ -1,27 +1,120 @@
+import com.azure.storage.blob.BlobServiceClientBuilder
+import com.azure.storage.common.StorageSharedKeyCredential
+import com.azure.storage.common.sas.{AccountSasPermission, AccountSasResourceType, AccountSasService, AccountSasSignatureValues}
 import com.typesafe.config.Config
 import cromwell.core.WorkflowOptions
 import cromwell.core.path.PathBuilderFactory
-import cromwell.filesystems.blob.BlobPathBuilder
 import net.ceedubs.ficus.Ficus._
 
+import java.time.OffsetDateTime
 import scala.concurrent.ExecutionContext
 import scala.concurrent.Future
 
+import scala.jdk.CollectionConverters._
+
 final case class BlobFileSystemConfig(config: Config)
 final case class BlobPathBuilderFactory(globalConfig: Config, instanceConfig: Config, singletonConfig: BlobFileSystemConfig) extends PathBuilderFactory {
   val sasToken: String = instanceConfig.as[String]("sas-token")
   val container: String = instanceConfig.as[String]("store")
   val endpoint: String = instanceConfig.as[String]("endpoint")
   val workspaceId: String = instanceConfig.as[String]("workspace-id")
-  val workspaceManagerURL = singletonConfig.config.as[String]("workspace-manager-url")
+  val workspaceManagerURL: String = singletonConfig.config.as[String]("workspace-manager-url")
+
+  val blobTokenGenerator: BlobTokenGenerator = BlobTokenGenerator.createBlobTokenGenerator(
+    container, endpoint, Option(workspaceId), Option(workspaceManagerURL))
 
   override def withOptions(options: WorkflowOptions)(implicit as: ActorSystem, ec: ExecutionContext): Future[BlobPathBuilder] = {
     Future {
-      new BlobPathBuilder(new AzureSasCredential(sasToken), container, endpoint)
+      new BlobPathBuilder(blobTokenGenerator, container, endpoint)
+    }
+  }
+}
+
+sealed trait BlobTokenGenerator {
+  def getAccessToken: String
+}
+
+object BlobTokenGenerator {
+  def createBlobTokenGenerator(container: String, endpoint: String): BlobTokenGenerator = {
+    createBlobTokenGenerator(container, endpoint, None, None)
+  }
+  def createBlobTokenGenerator(container: String, endpoint: String, workspaceId: Option[String], workspaceManagerURL: Option[String]): BlobTokenGenerator = {
+     (container: String, endpoint: String, workspaceId, workspaceManagerURL) match {
+       case (container, endpoint, None, None) =>
+         NativeBlobTokenGenerator(container, endpoint)
+       case (container, endpoint, Some(workspaceId), Some(workspaceManagerURL)) =>
+         WSMBlobTokenGenerator(container, endpoint, workspaceId, workspaceManagerURL)
+       case _ =>
+         throw new Exception("Bad! >:(")
+     }
+  }
+}
+
+case class WSMBlobTokenGenerator(container: String, endpoint: String, workspaceId: String, workspaceManagerURL: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
+    throw new NotImplementedError
+  }
+}
+
+case class NativeBlobTokenGenerator(container: String, endpoint: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
```

Comment:
as it is, this function is mostly a copy/paste job from BT-717. It'd be great to talk about best ways of abstracting some of this functionality -- the aspect of this that I'm least certain about is where each piece of functionality should live.

E.g. we could do something like:

```
def authenticatedAzureResourceManager() {
    val profile = new AzureProfile(AzureEnvironment.AZURE)
    val azureCredential = new DefaultAzureCredentialBuilder()
      .authorityHost(profile.getEnvironment.getActiveDirectoryEndpoint)
      .build
   AzureResourceManager.authenticate(azureCredential, profile).withDefaultSubscription
}
```

... but I'm not sure where something like that should live.

(Not necessarily endorsing the above abstraction, just using it to illustrate my uncertainty)

Reply:
Lets chat in mob time about this! I think making some helper methods to capture the broad steps for this would help the readability a bit, and they could go on the class or if they are static, on the companion object. The suggestion you had I thought was good, and possibly in addition you could try a method for the sharedStorageBlobCredential, and/or one for producing the permissions object.